### PR TITLE
Added RL vs DPC comparison for building control

### PR DESCRIPTION
**Pull request by Alirezad126**
Developed code to evaluate and compare the performance of the Primal-Dual DDPG agent against Differentiable Predictive Control (DPC) in building temperature regulation task.

### CODE REVIEW

**Comment by drgona** [comment]
Hi @Alirezad126, great notebook! Thank you very much for your contribution.

I only have a few minor suggestions for the notebook to be more didactical:

1. I suggest adding a short markdown description in the first cell summarizing what this notebook is about. You can include motivation, a brief summary of compared methods, a few references, and expected outcomes.
2. in cells where you use external dependencies that are a major part of the code, I suggest adding a brief description of the library and the class you are using. For instance, the gym environment. You can also include a link to a blog post or other resource that explains step by step how the user can customize the code and create their own gym library.
3. When using methods like replay buffer and PD-DDPG, let's add a markdown to briefly summarize what the method is about upfront; also, let's include references to papers or other resources where the user can learn more about these methods. 
4. I like the way you connected the code with markup in the `actor_loss `, it would be great to do more of this for the rest of the markup. So it is easier for the user to match the code with markup. For instance, Target Network Soft Updates markup is a bit harder to connect with the code. It would help to make this connection explicit in the text, also adding reference to this method would help to follow up for a curious user.
5. To structure the markup and code a bit better, I suggest avoiding large blocks of code bundled up in a single cell. For instance, you can have one cell with markup and one with code for each part of the RL method, i.e., to have actor, critic, and PD_DDPG class in separate cells preceded with markup linked with only that cell. 
6. Training Loop function for training the agent can also benefit from the markup to briefly summarize what is the code in the `training_loop `about. 
7. For the DPC part of the code, in the first cell markup you can also add references to papers and existing DPC examples in neuromancer.
8. Evaluate the RL and DPC models, which can include short markup to explain what metric you are using for comparison and why. Besides energy use and constraints violations, I suggest evaluating training time it took to train these policies.
9. We can wrap up with a cell including a short summary or conclusion about both methods their pros and cons and potential future research directions.

**Comment by RBirmiwal** [comment]
Hi @Alirezad126, also please change the destination branch to develop rather than master. Thank you

**Comment by Alirezad126** [comment]
> Hi @Alirezad126, also please change the destination branch to develop rather than master. Thank you

Hi @RBirmiwal . Changed the destination to develop branch.

**Comment by Alirezad126** [comment]
> Hi @Alirezad126, great notebook! Thank you very much for your contribution.
> 
> I only have a few minor suggestions for the notebook to be more didactical:
> 
>     1. I suggest adding a short markdown description in the first cell summarizing what this notebook is about. You can include motivation, a brief summary of compared methods, a few references, and expected outcomes.
> 
>     2. in cells where you use external dependencies that are a major part of the code, I suggest adding a brief description of the library and the class you are using. For instance, the gym environment. You can also include a link to a blog post or other resource that explains step by step how the user can customize the code and create their own gym library.
> 
>     3. When using methods like replay buffer and PD-DDPG, let's add a markdown to briefly summarize what the method is about upfront; also, let's include references to papers or other resources where the user can learn more about these methods.
> 
>     4. I like the way you connected the code with markup in the `actor_loss `, it would be great to do more of this for the rest of the markup. So it is easier for the user to match the code with markup. For instance, Target Network Soft Updates markup is a bit harder to connect with the code. It would help to make this connection explicit in the text, also adding reference to this method would help to follow up for a curious user.
> 
>     5. To structure the markup and code a bit better, I suggest avoiding large blocks of code bundled up in a single cell. For instance, you can have one cell with markup and one with code for each part of the RL method, i.e., to have actor, critic, and PD_DDPG class in separate cells preceded with markup linked with only that cell.
> 
>     6. Training Loop function for training the agent can also benefit from the markup to briefly summarize what is the code in the `training_loop `about.
> 
>     7. For the DPC part of the code, in the first cell markup you can also add references to papers and existing DPC examples in neuromancer.
> 
>     8. Evaluate the RL and DPC models, which can include short markup to explain what metric you are using for comparison and why. Besides energy use and constraints violations, I suggest evaluating training time it took to train these policies.
> 
>     9. We can wrap up with a cell including a short summary or conclusion about both methods their pros and cons and potential future research directions.

Hi @drgona. Thank you for your thoughtful comments. I will incorporate the requested modifications. Regarding the fair comparison, I need to ensure that the gym environment uses the exact same training and validation data as the DPC controller. Additionally, during the evaluation process, the policy should be assessed on the same dataset to maintain consistency in the loss values for comparison. To achieve a fair comparison of learning time, I will also introduce a specific threshold (such as a target loss value during evaluation) to determine when the learning process should stop.

**Comment by Alirezad126** [comment]
Hi @drgona,

I noticed an issue with the logic of the problem. Since the goal is to maintain the temperature within the specified reference bounds, but the HVAC system is only equipped to use heating energy to warm the building, certain scenarios can arise where the controller is forced to remain idle and select 0 as the action.

For instance, if the outdoor temperature and solar radiation are high, and the reference temperature is lower than the current indoor temperature, the agent has no way to cool the building. In these cases, it ends up receiving negative rewards (or constraint violations) simply for doing nothing—even though it has no other choice. (it doesn't affect the performance of the agent too much :sweat_smile: but its logically wrong  )

I think we should consider redefining the constraint violation criteria to better account for these situations.

**Comment by drgona** [comment]
Hi, @Alirezad126, yes, that's right. In this scenario when only heating is allowed, there will inevitably be constraint violations during hot day scenarios. I believe for tutorial purposes, the example is good as is, and the constraints don't require any special treatment.

**Comment by Alirezad126** [comment]
Hi @drgona, I have addressed all your comments on the first file. You can check and review the latest file in my repository. I will be happy to know if you have further comments.

**Comment by Alirezad126** [comment]
> Hi @drgona,
> 
> I noticed an issue with the logic of the problem. Since the goal is to maintain the temperature within the specified reference bounds, but the HVAC system is only equipped to use heating energy to warm the building, certain scenarios can arise where the controller is forced to remain idle and select 0 as the action.
> 
> For instance, if the outdoor temperature and solar radiation are high, and the reference temperature is lower than the current indoor temperature, the agent has no way to cool the building. In these cases, it ends up receiving negative rewards (or constraint violations) simply for doing nothing—even though it has no other choice. (it doesn't affect the performance of the agent too much 😅 but its logically wrong )
> 
> I think we should consider redefining the constraint violation criteria to better account for these situations.

**Comment by drgona** [comment]
@Alirezad126, excellent job on this example! This is one of the best examples we will have in the library. Well documented and discussed.

**Comment by Alirezad126** [comment]
> @Alirezad126, excellent job on this example! This is one of the best examples we will have in the library. Well documented and discussed.

Hi @drgona. Happy to hear that 🙂
